File: src/main/java/org/academy/api/client/gui/animation/StateListAnimator.java
package org.academy.api.client.gui.animation;

import org.jspecify.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class StateListAnimator {
    private final List<StateTuple> tuples = new ArrayList<>();
    @Nullable
    private Animator runningAnimator = null;
    @Nullable
    private Animator lastMatch = null;

    public void addState(int stateMask, Animator animator) {
        tuples.add(new StateTuple(stateMask, animator));
    }

    public void setState(int stateMask) {
        Animator match = null;

        for (var tuple : tuples) {
            if ((stateMask & tuple.mask) == tuple.mask) {
                match = tuple.animator;
                break;
            }
        }

        if (match == lastMatch) {
            return;
        }

        if (runningAnimator != null) {
            runningAnimator.cancel();
        }

        lastMatch = match;
        runningAnimator = match;

        if (match != null) {
            match.start();
        }
    }

    public void jumpToCurrentState() {
        if (runningAnimator != null) {
            runningAnimator.end();
        }
    }

    private record StateTuple(int mask, Animator animator) {
    }
}

File: src/main/kotlin/org/academy/api/client/gui/animation/StateListAnimator.kt
package org.academy.api.client.gui.animation

class StateListAnimator {
    private val tuples: MutableList<StateTuple> = ArrayList<StateTuple>()
    private var runningAnimator: Animator? = null
    private var lastMatch: Animator? = null

    fun addState(stateMask: Int, animator: Animator) {
        tuples.add(StateTuple(stateMask, animator))
    }

    fun setState(stateMask: Int) {
        var match: Animator? = null

        for (tuple in tuples) {
            if ((stateMask and tuple.mask) == tuple.mask) {
                match = tuple.animator
                break
            }
        }

        if (match === lastMatch) {
            return
        }

        if (runningAnimator != null) {
            runningAnimator!!.cancel()
        }

        lastMatch = match
        runningAnimator = match

        match?.start()
    }

    fun jumpToCurrentState() {
        if (runningAnimator != null) {
            runningAnimator!!.end()
        }
    }

    private data class StateTuple(val mask: Int, val animator: Animator)
}